### DISCUSSION TITLE
Structured Arrays

### DISCUSSION THREAD

**Discussion by lorentzenchr** [wjakob/nanobind#352] [Q&A] [answered]
Numpy supports [structured arrays](https://numpy.org/doc/stable/user/basics.rec.html#structured-arrays), e.g.
```python
HISTOGRAM_DTYPE = np.dtype([
    ('sum_gradients', np.float64),
    ('sum_hessians', np.float64),
    ('count', np.uint32),
])

histograms = np.empty(shape=(3, 8), dtype=HISTOGRAM_DTYPE)
```
In Cython, this then looks like
```cython
cdef packed struct hist_struct:
    # Same as histogram dtype but we need a struct to declare views. It needs
    # to be packed since by default numpy dtypes aren't aligned
    Y_DTYPE_C sum_gradients
    Y_DTYPE_C sum_hessians
    unsigned int count

hist_struct [:, ::1] histograms_view = np.empty(shape=(3, 8), dtype=HISTOGRAM_DTYPE)
```

**Is it possible to achieve the same with nanobind?**

Note: Examples taken from scikit-learn histograms in the `HistGradientBoostingRegressor`.

**Comment by wjakob** [comment, marked as answer]
No, and the documentation goes into more detail: https://nanobind.readthedocs.io/en/latest/ndarray.html#limitations-related-to-dtypes

**Comment by lorentzenchr** [comment]
Wow, thanks for the suuuuper fast reply!
So if one needs those structures, say just internally, one would need to keep it in C++ only and, if necessary, handle (lifetime) binding to Python oneself, right? (Maybe I use the wrong terms, but I hope you know what I mean)
